### ISSUE TITLE
[Quality] Add Error Message Context and Suggestions

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2641]
## Problem

Error messages may be too generic, lacking context about what went wrong and how to fix it.

## Current Issues

**Generic errors:**
```mojo
raise Error("Invalid shape")
```

**Better errors:**
```mojo
raise Error(
    "Invalid shape for matrix multiplication: " +
    "Cannot multiply shapes " + str(a.shape()) + " and " + str(b.shape()) + ". " +
    "Hint: Ensure inner dimensions match (got " + str(a.shape()[1]) + " != " + str(b.shape()[0]) + ")"
)
```

## Guidelines for Better Errors

1. **Include context**: What operation failed
2. **Show values**: Display actual values that caused the error
3. **Suggest fix**: Hint at how to resolve
4. **Be specific**: Avoid generic "invalid input"

## Examples

```mojo
# Broadcasting error
raise Error(
    "Cannot broadcast shapes " + str(a.shape()) + " and " + str(b.shape()) + ". " +
    "Shapes must be compatible for broadcasting. " +
    "See: docs/broadcasting.md"
)

# Type error
raise Error(
    "Expected tensor dtype " + str(expected_dtype) + " but got " + str(actual_dtype) + ". " +
    "Use tensor.cast(" + str(expected_dtype) + ") to convert."
)

# Shape error
raise Error(
    "Expected 4D tensor (N, C, H, W) but got " + str(tensor.dim()) + "D tensor with shape " + str(tensor.shape()) + ". " +
    "Hint: Use tensor.reshape() or add batch/channel dimensions."
)
```

## Impact

- **Severity**: LOW
- **Developer experience**: Faster debugging
- **Learning curve**: Easier for beginners

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Limit reached · resets 2pm (America/Los_Angeles)

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Now I have all the information I need. Let me create a comprehensive implementation plan:

---

# Implementation Plan: Issue #2641 - Add Error Message Context and Suggestions

## 1. Summary

This plan addresses the quality improvement issue of enhancing error messages throughout the ML Odyssey codebase. The goal is to transform generic error messages into context-rich, actionable messages that include:
- **What operation failed** (function/operation name)
- **Actual values** that caused the error
- **Expected values/constraints**
- **Suggestions** for how to fix the issue

### Current State Analysis

Based on codebase exploration:
- **488 total error raises** across the codebase
- **~63 short/generic error messages** (under 30 characters) that need improvement
- **Excellent patterns exist** in `shared/core/validation.mojo` with helper functions
- **Inconsistent error quality** across modules - some have great messages, others are generic

### Priority Modules (by error count and criticality)

| Module | Error Count | Priority | Reason |
|--------|------------|----------|--------|
| `shared/core/extensor.mojo` | 50+ | HIGH | Core tensor operations, most common |
| `shared/core/shape.mojo` | 20+ | HIGH | Shape manipulation, frequent errors |
| `shared/core/matmul.mojo` | 10+ | HIGH | Matrix multiplication, critical path |
| `shared/core/conv.mojo` | 15+ | HIGH | Convolution, complex shapes |
| `shared/core/broadcasting.mojo` | 5+ | HIGH | Broadcasting rules, confusing errors |
| `shared/core/validation.mojo` | 15+ | LOW | Already has best practices (reference) |
| `shared/utils/config.mojo` | 25+ | MEDIUM | Config loading, user-facing |
| `shared/training/` | 30+ | MEDIUM | Training utilities |

---

## 2. Step-by-Step Implementation Tasks

### Phase 1: Create Error Utilities Module (Foundation)

**Task 1.1: Create `shared/core/error_utils.mojo`**

Create a centralized error utilities module that expands on the existing patterns in `validation.mojo`:

```mojo
"""Error formatting utilities for consistent, context-rich error messages.

Provides helper functions and constants for creating descriptive error messages
that include operation context, actual vs expected values, and fix suggestions.
"""

# Error message template constants
alias ERROR_TEMPLATE_SHAPE = "{op}: expected shape {expected}, got shape {actual}"
alias ERROR_TEMPLATE_DTYPE = "{op}: expected dtype {expected}, got dtype {actual}"
alias ERROR_TEMPLATE_DIM = "{op}: expected {expected}D tensor, got {expected}D with shape {shape}"

# Helper functions
fn format_shape(shape: List[Int]) -> String
fn format_dtype(dtype: DType) -> String
fn format_dim_mismatch(op: String, expected_dim: Int, actual_dim: Int, actual_shape: List[Int]) -> String
fn format_dtype_mismatch(op: String, expected: DType, actual: DType) -> String
fn format_shape_mismatch(op: String, expected: List[Int], actual: List[Int]) -> String
fn format_matmul_error(a_shape: List[Int], b_shape: List[Int]) -> String
fn format_broadcast_error(shape1: List[Int], shape2: List[Int], dim: Int) -> String
```

**Task 1.2: Add fix suggestion helpers**

```mojo
fn suggest_reshape(current_shape: List[Int], target_shape: List[Int]) -> String
fn suggest_unsqueeze(shape: List[Int], missing_dim: String) -> String
fn suggest_cast(current_dtype: DType, target_dtype: DType) -> String
fn suggest_transpose(shape: List[Int]) -> String
```

### Phase 2: Update Core Tensor Operations (Highest Priority)

**Task 2.1: Update `shared/core/extensor.mojo`**

Current (line 552):
```mojo
raise Error("Cannot reshape: element count mismatch")
```

Improved:
```mojo
raise Error(
    "reshape: cannot reshape tensor with " + String(self.numel()) + 
    " elements to shape " + format_shape(new_shape) + 
    " (" + String(new_numel) + " elements). " +
    "Hint: Ensure total elements match (current shape: " + format_shape(self._shape) + ")"
)
```

**Files to update in extensor.mojo:**
- Line 139: Tensor size validation
- Line 312: Index bounds checking  
- Line 366: Shape validation
- Line 552: Reshape element mismatch
- Line 604-623: Slice bounds checking
- Line 678: Index out of bounds
- Lines 1162-1731: FP8/BF8/MX4/NVFP4 conversion errors

**Task 2.2: Update `shared/core/shape.mojo`**

Current (line 146):
```mojo
raise Error("view: new shape must have same number of elements")
```

Improved:
```mojo
raise Error(
    "view: cannot create view with " + String(new_numel) + " elements " +
    "from tensor with " + String(old_numel) + " elements. " +
    "Hint: Use reshape() if element counts differ, or ensure shapes are compatible"
)
```

**Files to update:**
- Lines 146, 184-219: Reshape validation
- Lines 263-266: Squeeze validation
- Line 323: Unsqueeze validation
- Lines 437-468: Concatenate validation
- Lines 535-562: Stack validation
- Line 724: Flatten validation

**Task 2.3: Update `shared/core/matmul.mojo`**

Current (line 96-101):
```mojo
raise Error(
    "matmul_typed: incompatible dimensions for matmul: "
    + String(K) + " != " + String(b_shape[0])
)
```

Improved:
```mojo
raise Error(
    "matmul: cannot multiply matrices A" + format_shape(a_shape) + 
    " and B" + format_shape(b_shape) + ". " +
    "Inner dimensions must match: A has " + String(K) + " columns, " +
    "B has " + String(b_shape[0]) + " rows. " +
    "Hint: B should have shape [" + String(K) + ", N] for any N"
)
```

### Phase 3: Update Neural Network Operations

**Task 3.1: Update `shared/core/conv.mojo`**

Current (line 83):
```mojo
raise Error("Input must be 4D tensor (batch, channels, height, width)")
```

Improved:
```mojo
raise Error(
    "conv2d: expected 4D input tensor (batch, channels, height, width), " +
    "got " + String(len(x_shape)) + "D tensor with shape " + format_shape(x_shape) + ". " +
    "Hint: Use input.unsqueeze(0) to add batch dimension or reshape to 4D"
)
```

**Task 3.2: Update `shared/core/broadcasting.mojo`**

Current (line 50-57):
```mojo
raise Error(
    "Shapes are not broadcast-compatible: dimension " + String(i) + 
    " has sizes " + String(dim1) + " and " + String(dim2)
)
```

Improved:
```mojo
raise Error(
    "broadcast: shapes " + format_shape(shape1) + " and " + format_shape(shape2) + 
    " are not broadcast-compatible. " +
    "At dimension " + String(max_ndim - 1 - i) + " (from right): " +
    "got sizes " + String(dim1) + " and " + String(dim2) + ". " +
    "Hint: Broadcasting requires dimensions to be equal or one to be 1. " +
    "See: https://numpy.org/doc/stable/user/basics.broadcasting.html"
)
```

**Task 3.3: Update `shared/core/loss.mojo`**

Current (lines 68-71):
```mojo
raise Error("Predictions and targets must have the same dtype")
raise Error("Predictions and targets must have the same shape")
```

Improved:
```mojo
raise Error(
    "binary_cross_entropy: dtype mismatch - predictions have dtype " + 
    format_dtype(predictions.dtype()) + " but targets have dtype " + 
    format_dtype(targets.dtype()) + ". " +
    "Hint: Use tensor.cast(" + format_dtype(targets.dtype()) + ") to convert"
)

raise Error(
    "binary_cross_entropy: shape mismatch - predictions have shape " +
    format_shape(predictions.shape()) + " but targets have shape " +
    format_shape(targets.shape()) + ". " +
    "Hint: Ensure both tensors have identical shapes"
)
```

**Task 3.4: Update `shared/core/pooling.mojo`**

Current (line 60):
```mojo
raise Error("Input must be 4D tensor (batch, channels, height, width)")
```

Improved:
```mojo
raise Error(
    "maxpool2d: expected 4D input tensor (batch, channels, height, width), " +
    "got " + String(len(x_shape)) + "D tensor with shape " + format_shape(x_shape) + ". " +
    "Hint: Use input.unsqueeze(0) for batch dimension, input.unsqueeze(1) for channel dimension"
)
```

### Phase 4: Update Training and Utilities

**Task 4.1: Update `shared/utils/config.mojo`**

Current (line 545):
```mojo
raise Error("Key not found: " + key)
```

Improved:
```mojo
raise Error(
    "Config key '" + key + "' not found. " +
    "Available keys: " + format_available_keys(self) + ". " +
    "Hint: Check config file for typos or use get_or_default()"
)
```

**Task 4.2: Update `shared/training/optimizers/*.mojo`**

Current (line 79-82 in sgd.mojo):
```mojo
raise Error("Parameters and gradients must have the same shape")
raise Error("Parameters and gradients must have the same dtype")
```

Improved:
```mojo
raise Error(
    "SGD.step: parameter shape " + format_shape(params.shape()) + 
    " doesn't match gradient shape " + format_shape(grads.shape()) + ". " +
    "Hint: Ensure gradients were computed for the correct parameters"
)
```

### Phase 5: Update Data Loading Operations

**Task 5.1: Update `shared/data/loaders.mojo`**

**Task 5.2: Update `shared/data/transforms.mojo`**

**Task 5.3: Update `shared/data/formats/idx_loader.mojo`**

---

## 3. Files to Modify/Create

### New Files

| File | Purpose |
|------|---------|
| `shared/core/error_utils.mojo` | Centralized error formatting utilities |
| `tests/shared/core/test_error_utils.mojo` | Unit tests for error utilities |
| `docs/adr/ADR-XXX-error-message-guidelines.md` | ADR documenting error message standards |

### Files to Modify

**Phase 2 (Core - HIGH priority):**
- `shared/core/extensor.mojo` (~50 error messages)
- `shared/core/shape.mojo` (~20 error messages)
- `shared/core/matmul.mojo` (~10 error messages)
- `shared/core/broadcasting.mojo` (~5 error messages)
- `shared/core/arithmetic.mojo` (~10 error messages)

**Phase 3 (Neural Network - HIGH priority):**
- `shared/core/conv.mojo` (~15 error messages)
- `shared/core/pooling.mojo` (~8 error messages)
- `shared/core/loss.mojo` (~20 error messages)
- `shared/core/activation.mojo` (~25 error messages)
- `shared/core/normalization.mojo` (~15 error messages)

**Phase 4 (Training - MEDIUM priority):**
- `shared/utils/config.mojo` (~25 error messages)
- `shared/training/optimizers/sgd.mojo` (~5 error messages)
- `shared/training/optimizers/adam.mojo` (~5 error messages)
- `shared/training/metrics/*.mojo` (~10 error messages)

**Phase 5 (Data - MEDIUM priority):**
- `shared/data/loaders.mojo` (~10 error messages)
- `shared/data/transforms.mojo` (~10 error messages)
- `shared/data/formats/idx_loader.mojo` (~8 error messages)

---

## 4. Testing Approach

### Unit Tests for Error Utilities

```mojo
# tests/shared/core/test_error_utils.mojo

fn test_format_shape():
    var shape = List[Int]()
    shape.append(2)
    shape.append(3)
    shape.append(4)
    assert_equal(format_shape(shape), "[2, 3, 4]")

fn test_format_matmul_error():
    var a_shape = List[Int]()
    a_shape.append(32)
    a_shape.append(64)
    var b_shape = List[Int]()
    b_shape.append(128)
    b_shape.append(32)
    
    var error_msg = format_matmul_error(a_shape, b_shape)
    assert_true("64" in error_msg)  # Shows A's columns
    assert_true("128" in error_msg)  # Shows B's rows
    assert_true("Hint" in error_msg)  # Has suggestion
```

### Integration Tests

Test that improved error messages appear when expected:

```mojo
fn test_matmul_error_message_quality():
    """Verify matmul produces descriptive error on dimension mismatch."""
    var a = zeros([32, 64], DType.float32)
    var b = zeros([128, 32], DType.float32)  # Wrong shape
    
    try:
        var c = matmul(a, b)
        raise Error("Should have raised")
    except e:
        var msg = String(e)
        assert_true("matmul" in msg, "Error should mention operation")
        assert_true("[32, 64]" in msg, "Error should show A shape")
        assert_true("[128, 32]" in msg, "Error should show B shape")
        assert_true("Hint" in msg, "Error should include suggestion")
```

### Error Message Quality Checklist

Each improved error should be tested to verify:
- [ ] Includes operation name
- [ ] Shows actual values
- [ ] Shows expected values (where applicable)
- [ ] Includes actionable hint/suggestion
- [ ] Is under 200 characters for readability

---

## 5. Success Criteria

### Quantitative Metrics

| Metric | Current | Target |
|--------|---------|--------|
| Generic errors (<30 chars) | 63 | <10 |
| Errors with operation name | ~60% | 100% |
| Errors with actual values | ~40% | 95% |
| Errors with suggestions | ~5% | 80% |

### Qualitative Criteria

1. **Consistency**: All error messages follow the standard format:
   ```
   "{operation}: {what_failed} - {actual_vs_expected}. Hint: {suggestion}"
   ```

2. **Actionability**: Every error message should help users understand:
   - What went wrong
   - Why it went wrong
   - How to fix it

3. **Documentation**: ADR created documenting error message guidelines

### Validation

1. All existing tests continue to pass
2. New tests for error utilities pass
3. Code review confirms error message quality
4. Pre-commit hooks pass (formatting, linting)

---

## 6. Implementation Schedule

| Phase | Files | Estimated Effort | Dependencies |
|-------|-------|------------------|--------------|
| Phase 1 | error_utils.mojo | 1 day | None |
| Phase 2 | extensor, shape, matmul, broadcasting, arithmetic | 2 days | Phase 1 |
| Phase 3 | conv, pooling, loss, activation, normalization | 2 days | Phase 1 |
| Phase 4 | config, optimizers, metrics | 1 day | Phase 1 |
| Phase 5 | loaders, transforms, idx_loader | 1 day | Phase 1 |
| Testing | Unit and integration tests | 1 day | All phases |
| Documentation | ADR, update CLAUDE.md | 0.5 day | All phases |

**Total Estimated Effort**: 8.5 days

---

## 7. Risk Assessment

| Risk | Impact | Mitigation |
|------|--------|------------|
| Error message changes break existing tests | Medium | Run full test suite after each phase |
| String concatenation performance impact | Low | Errors are exceptional paths, not hot paths |
| Scope creep (too many files) | Medium | Prioritize HIGH priority modules first |
| Inconsistent application | Medium | Create ADR with clear guidelines |

---

## 8. Example Before/After

### Matrix Multiplication

**Before:**
```
Error: matmul: requires 2D tensors
```

**After:**
```
Error: matmul: expected 2D tensors for matrix multiplication.
  Got: A is 3D with shape [32, 64, 128], B is 2D with shape [128, 64].
  Hint: Use A.reshape([batch * seq, hidden]) to flatten A to 2D, 
        or use batch_matmul() for batched matrix multiplication.
```

### Shape Mismatch

**Before:**
```
Error: Predictions and targets must have the same shape
```

**After:**
```
Error: cross_entropy: shape mismatch between predictions and targets.
  Predictions: [32, 10] (batch=32, classes=10)
  Targets: [32] (batch=32)
  Hint: Targets should be [32] for class indices or [32, 10] for one-hot encoding.
        Current targets appear to be class indices (expected format).
```

### Broadcasting

**Before:**
```
Error: Shapes are not broadcast-compatible: dimension 0 has sizes 3 and 2
```

**After:**
```
Error: add: cannot broadcast shapes [3, 4, 5] and [2, 4, 5].
  Dimension 0 (leftmost): sizes 3 and 2 are incompatible.
  Broadcasting requires dimensions to be equal or one to be 1.
  Hint: Reshape one tensor to have compatible dimensions, or use 
        explicit broadcasting with expand().
  See: docs/broadcasting.md
```

---

## Resource Usage
- Tool calls used: 20/50
- Steps used: 20/50

**Comment by mvillmow** [comment]
## Starting Implementation

I'm beginning implementation of error message context and suggestions following the detailed plan.

### Approach
1. Create  with centralized error formatting utilities
2. Update core tensor operations (extensor, shape, matmul, broadcasting)
3. Add comprehensive tests
4. Create ADR documenting error message guidelines

### Implementation Progress
- [ ] Phase 1: Create error utilities module
- [ ] Phase 2: Update core tensor operations  
- [ ] Phase 3: Add tests
- [ ] Phase 4: Create ADR
- [ ] Phase 5: Create PR

Working autonomously...

**Comment by mvillmow** [comment]
## Implementation Progress Update

### Completed
- ✅ Created `shared/core/error_utils.mojo` with comprehensive error formatting utilities including:
  - `format_shape()` - Convert shape to bracketed string
  - `format_dtype()` - Convert dtype to readable string
  - `format_matmul_error()` - Specialized matmul error messages
  - `format_broadcast_error()` - Broadcasting failure messages
  - `format_reshape_error()` - Reshape element mismatch messages
  - Suggestion helpers for common fixes

### In Progress
- Working on updating `shared/core/matmul.mojo` with improved error messages
- Encountered file locking issue - investigating

### Next Steps
1. Complete matmul.mojo updates
2. Create comprehensive tests for error_utils.mojo
3. Update additional core files (shape.mojo, broadcasting.mojo)
4. Create ADR documenting error message guidelines

Continuing work...
